Use Cmdline from linux_loader instead of String

Figure out the most straightforward way to reuse `Cmdline` from the `linux-loader` crate instead of `String` for building/passing/handling parts of the kernel command line that we're passing to the guest (also double check there are no concerns or missing features beforehand).

Hey friends I tried to build after adding linux-loader to Cargo.toml & running `cargo build`

```
error[E0432]: unresolved import `vmm_sys_util::eventfd`
  --> /Users/ant/.cargo/registry/src/github.com-1ecc6299db9ec823/vm-superio-0.1.1/src/serial.rs:15:19
   |
15 | use vmm_sys_util::eventfd::EventFd;
   |                   ^^^^^^^ could not find `eventfd` in `vmm_sys_util`

error[E0432]: unresolved import `vmm_sys_util::epoll`
  --> /Users/ant/.cargo/registry/src/github.com-1ecc6299db9ec823/event-manager-0.2.1/src/lib.rs:16:23
   |
16 | pub use vmm_sys_util::epoll::EventSet;
   |                       ^^^^^ could not find `epoll` in `vmm_sys_util`

error[E0432]: unresolved import `vmm_sys_util::epoll`
 --> /Users/ant/.cargo/registry/src/github.com-1ecc6299db9ec823/event-manager-0.2.1/src/epoll.rs:7:19
  |
7 | use vmm_sys_util::epoll::{ControlOperation, Epoll, EpollEvent};
  |                   ^^^^^ could not find `epoll` in `vmm_sys_util`

error[E0432]: unresolved import `vmm_sys_util::epoll`
 --> /Users/ant/.cargo/registry/src/github.com-1ecc6299db9ec823/event-manager-0.2.1/src/events.rs:8:19
  |
8 | use vmm_sys_util::epoll::{ControlOperation, EpollEvent, EventSet};
  |                   ^^^^^ could not find `epoll` in `vmm_sys_util`

error[E0432]: unresolved import `vmm_sys_util::epoll`
 --> /Users/ant/.cargo/registry/src/github.com-1ecc6299db9ec823/event-manager-0.2.1/src/manager.rs:6:19
  |
6 | use vmm_sys_util::epoll::EpollEvent;
  |                   ^^^^^ could not find `epoll` in `vmm_sys_util`

error[E0432]: unresolved import `vmm_sys_util::epoll`
 --> /Users/ant/.cargo/registry/src/github.com-1ecc6299db9ec823/event-manager-0.2.1/src/manager.rs:8:19
  |
8 | use vmm_sys_util::epoll::{ControlOperation, EventSet};
  |                   ^^^^^ could not find `epoll` in `vmm_sys_util`

error[E0432]: unresolved import `vmm_sys_util::eventfd`
  --> /Users/ant/.cargo/registry/src/github.com-1ecc6299db9ec823/event-manager-0.2.1/src/endpoint.rs:26:19
   |
26 | use vmm_sys_util::eventfd::{EventFd, EFD_NONBLOCK};
   |                   ^^^^^^^ could not find `eventfd` in `vmm_sys_util`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: could not compile `vm-superio`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0432`.
error: build failed
```

I'm on a macos but i feel like that shouldnt matter wrt cargo build ...

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,3 +19,6 @@ codegen-units = 1
 lto = true
 panic = "abort"
 
+[patch.crates-io]
+# TODO: Using this patch until a version > 4.0 gets published.
+linux-loader = { git = "https://github.com/rust-vmm/linux-loader.git", rev = "9a9f071" }

--- a/coverage_config_x86_64.json
+++ b/coverage_config_x86_64.json
@@ -1,5 +1,5 @@
 {
-  "coverage_score": 70.2,
+  "coverage_score": 70.4,
   "exclude_path": "msr_index.rs,mpspec.rs,tests/,src/devices/src/virtio/net/bindings.rs",
   "crate_features": ""
 }

--- a/src/api/Cargo.toml
+++ b/src/api/Cargo.toml
@@ -8,3 +8,6 @@ edition = "2018"
 clap = "2.33.3"
 
 vmm = { path = "../vmm" }
+
+[dev-dependencies]
+linux-loader = "0.4.0"

--- a/src/api/src/lib.rs
+++ b/src/api/src/lib.rs
@@ -82,7 +82,9 @@ mod tests {
     use super::*;
 
     use std::path::PathBuf;
-    use vmm::DEFAULT_KERNEL_CMDLINE;
+
+    use linux_loader::cmdline::Cmdline;
+
     use vmm::{KernelConfig, MemoryConfig, VcpuConfig};
 
     #[test]
@@ -212,6 +214,9 @@ mod tests {
         ])
         .is_err());
 
+        let mut foo_cmdline = Cmdline::new(4096);
+        foo_cmdline.insert_str("\"foo=bar bar=foo\"").unwrap();
+
         // OK.
         assert_eq!(
             Cli::launch(vec![
@@ -227,7 +232,7 @@ mod tests {
             VMMConfig {
                 kernel_config: KernelConfig {
                     path: PathBuf::from("/foo/bar"),
-                    cmdline: "\"foo=bar bar=foo\"".to_string(),
+                    cmdline: foo_cmdline,
                     load_addr: 42,
                 },
                 memory_config: MemoryConfig { size_mib: 128 },
@@ -243,7 +248,7 @@ mod tests {
             VMMConfig {
                 kernel_config: KernelConfig {
                     path: PathBuf::from("/foo/bar"),
-                    cmdline: DEFAULT_KERNEL_CMDLINE.to_string(),
+                    cmdline: KernelConfig::default_cmdline(),
                     load_addr: 1048576,
                 },
                 memory_config: MemoryConfig { size_mib: 256 },

--- a/src/devices/Cargo.toml
+++ b/src/devices/Cargo.toml
@@ -11,7 +11,7 @@ kvm-ioctls = "0.8.0"
 libc = "0.2.76"
 linux-loader = "0.3.0"
 log = "0.4.6"
-vm-memory = "0.5.0"
+vm-memory = "0.6.0"
 vm-superio = "0.4.0"
 vmm-sys-util = "0.8.0"
 
@@ -27,4 +27,4 @@ virtio-queue = { git = "https://github.com/rust-vmm/vm-virtio.git"}
 utils = { path = "../utils" }
 
 [dev-dependencies]
-vm-memory = { version = "0.5.0", features = ["backend-mmap"] }
+vm-memory = { version = "0.6.0", features = ["backend-mmap"] }

--- a/src/devices/Cargo.toml
+++ b/src/devices/Cargo.toml
@@ -9,7 +9,7 @@ license = "Apache-2.0 OR BSD-3-Clause"
 event-manager = { version = "0.2.1", features = ["remote_endpoint"] }
 kvm-ioctls = "0.8.0"
 libc = "0.2.76"
-linux-loader = "0.3.0"
+linux-loader = "0.4.0"
 log = "0.4.6"
 vm-memory = "0.6.0"
 vm-superio = "0.4.0"

--- a/src/vm-vcpu/Cargo.toml
+++ b/src/vm-vcpu/Cargo.toml
@@ -10,7 +10,7 @@ edition = "2018"
 libc = "0.2.76"
 kvm-bindings = { version = "0.4.0", features = ["fam-wrappers"] }
 kvm-ioctls = "0.8.0"
-vm-memory = "0.5.0"
+vm-memory = "0.6.0"
 vmm-sys-util = "0.8.0"
 
 utils = { path = "../utils" }
@@ -21,4 +21,4 @@ utils = { path = "../utils" }
 vm-device = { git = "https://github.com/rust-vmm/vm-device", rev = "5847f12" }
 
 [dev-dependencies]
-vm-memory = { version = "0.5.0", features = ["backend-mmap"] }
+vm-memory = { version = "0.6.0", features = ["backend-mmap"] }

--- a/src/vm-vcpu/src/vcpu/mptable.rs
+++ b/src/vm-vcpu/src/vcpu/mptable.rs
@@ -284,7 +284,9 @@ pub fn setup_mptable<M: GuestMemory>(mem: &M, num_cpus: u8) -> Result<()> {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use vm_memory::{Bytes, GuestMemoryMmap};
+
+    use vm_memory::{self, Bytes};
+    type GuestMemoryMmap = vm_memory::GuestMemoryMmap<()>;
 
     fn table_entry_size(type_: u8) -> usize {
         match u32::from(type_) {

--- a/src/vm-vcpu/src/vm.rs
+++ b/src/vm-vcpu/src/vm.rs
@@ -131,26 +131,24 @@ impl<EH: 'static + ExitHandler + Send> KvmVm<EH> {
         }
 
         // Register guest memory regions with KVM.
-        guest_memory
-            .with_regions(|index, region| {
-                let memory_region = kvm_userspace_memory_region {
-                    slot: index as u32,
-                    guest_phys_addr: region.start_addr().raw_value(),
-                    memory_size: region.len() as u64,
-                    // It's safe to unwrap because the guest address is valid.
-                    userspace_addr: guest_memory.get_host_address(region.start_addr()).unwrap()
-                        as u64,
-                    flags: 0,
-                };
+        for (index, region) in guest_memory.iter().enumerate() {
+            let memory_region = kvm_userspace_memory_region {
+                slot: index as u32,
+                guest_phys_addr: region.start_addr().raw_value(),
+                memory_size: region.len() as u64,
+                // It's safe to unwrap because the guest address is valid.
+                userspace_addr: guest_memory.get_host_address(region.start_addr()).unwrap() as u64,
+                flags: 0,
+            };
 
-                // Safe because:
-                // * userspace_addr is a valid address for a memory region, obtained by calling
-                //   get_host_address() on a valid region's start address;
-                // * the memory regions do not overlap - there's either a single region spanning
-                //   the whole guest memory, or 2 regions with the MMIO gap in between.
-                unsafe { self.fd.set_user_memory_region(memory_region) }
-            })
-            .map_err(Error::SetupMemoryRegion)?;
+            // Safe because:
+            // * userspace_addr is a valid address for a memory region, obtained by calling
+            //   get_host_address() on a valid region's start address;
+            // * the memory regions do not overlap - there's either a single region spanning
+            //   the whole guest memory, or 2 regions with the MMIO gap in between.
+            unsafe { self.fd.set_user_memory_region(memory_region) }
+                .map_err(Error::SetupMemoryRegion)?;
+        }
 
         Ok(())
     }
@@ -266,7 +264,9 @@ mod tests {
     use std::time::Duration;
 
     use kvm_ioctls::Kvm;
-    use vm_memory::{Bytes, GuestAddress, GuestMemoryMmap};
+    use vm_memory::{Bytes, GuestAddress};
+
+    type GuestMemoryMmap = vm_memory::GuestMemoryMmap<()>;
 
     #[derive(Clone, Default)]
     struct WrappedExitHandler(Arc<DummyExitHandler>);

--- a/src/vmm/Cargo.toml
+++ b/src/vmm/Cargo.toml
@@ -10,7 +10,7 @@ event-manager = "0.2.1"
 kvm-bindings = { version = "0.4.0", features = ["fam-wrappers"] }
 kvm-ioctls = "0.8.0"
 libc = "0.2.91"
-linux-loader = { version = "0.3.0", features = ["bzimage", "elf"] }
+linux-loader = { version = "0.4.0", features = ["bzimage", "elf"] }
 vm-memory = { version = "0.6.0", features = ["backend-mmap"] }
 vm-superio = "0.4.0"
 vmm-sys-util = "0.8.0"

--- a/src/vmm/Cargo.toml
+++ b/src/vmm/Cargo.toml
@@ -11,7 +11,7 @@ kvm-bindings = { version = "0.4.0", features = ["fam-wrappers"] }
 kvm-ioctls = "0.8.0"
 libc = "0.2.91"
 linux-loader = { version = "0.3.0", features = ["bzimage", "elf"] }
-vm-memory = { version = "0.5.0", features = ["backend-mmap"] }
+vm-memory = { version = "0.6.0", features = ["backend-mmap"] }
 vm-superio = "0.4.0"
 vmm-sys-util = "0.8.0"
 

--- a/src/vmm/src/config/builder.rs
+++ b/src/vmm/src/config/builder.rs
@@ -192,7 +192,7 @@ mod tests {
     use std::path::PathBuf;
 
     use super::*;
-    use crate::{DEFAULT_HIGH_RAM_START, DEFAULT_KERNEL_CMDLINE};
+    use crate::DEFAULT_HIGH_RAM_START;
 
     #[test]
     fn test_builder_default_err() {
@@ -255,7 +255,7 @@ mod tests {
         assert_eq!(
             vmm_config.unwrap().kernel_config,
             KernelConfig {
-                cmdline: DEFAULT_KERNEL_CMDLINE.to_string(),
+                cmdline: KernelConfig::default_cmdline(),
                 load_addr: DEFAULT_HIGH_RAM_START,
                 path: PathBuf::from("bzImage")
             }
@@ -337,7 +337,7 @@ mod tests {
                 memory_config: MemoryConfig { size_mib: 1024 },
                 vcpu_config: VcpuConfig { num: 2 },
                 kernel_config: KernelConfig {
-                    cmdline: DEFAULT_KERNEL_CMDLINE.to_string(),
+                    cmdline: KernelConfig::default_cmdline(),
                     load_addr: DEFAULT_HIGH_RAM_START,
                     path: PathBuf::from("bzImage")
                 },

--- a/src/vmm/src/config/mod.rs
+++ b/src/vmm/src/config/mod.rs
@@ -7,6 +7,8 @@ use std::num;
 use std::path::PathBuf;
 use std::result;
 
+use linux_loader::cmdline::Cmdline;
+
 use arg_parser::CfgArgParser;
 use builder::Builder;
 
@@ -14,6 +16,8 @@ use super::{DEFAULT_KERNEL_CMDLINE, DEFAULT_KERNEL_LOAD_ADDR};
 
 mod arg_parser;
 mod builder;
+
+const KERNEL_CMDLINE_CAPACITY: usize = 4096;
 
 /// Errors encountered converting the `*Config` objects.
 #[derive(Clone, Debug, PartialEq)]
@@ -133,17 +137,30 @@ impl TryFrom<&str> for VcpuConfig {
 #[derive(Clone, Debug, PartialEq)]
 pub struct KernelConfig {
     /// Kernel command line.
-    pub cmdline: String,
+    pub cmdline: Cmdline,
     /// Path to the kernel image.
     pub path: PathBuf,
     /// Address where the kernel is loaded.
     pub load_addr: u64,
 }
 
+impl KernelConfig {
+    /// Return the default kernel command line used by the Vmm.
+    pub fn default_cmdline() -> Cmdline {
+        let mut cmdline = Cmdline::new(KERNEL_CMDLINE_CAPACITY);
+
+        // It's ok to use `unwrap` because the initial capacity of `cmdline` should be
+        // sufficient to accommodate the default kernel cmdline.
+        cmdline.insert_str(DEFAULT_KERNEL_CMDLINE).unwrap();
+
+        cmdline
+    }
+}
+
 impl Default for KernelConfig {
     fn default() -> Self {
         KernelConfig {
-            cmdline: String::from(DEFAULT_KERNEL_CMDLINE),
+            cmdline: KernelConfig::default_cmdline(),
             path: PathBuf::from(""), // FIXME: make it a const.
             load_addr: DEFAULT_KERNEL_LOAD_ADDR,
         }
@@ -158,10 +175,16 @@ impl TryFrom<&str> for KernelConfig {
         // `cmdline=<"string">,path=/path/to/kernel,kernel_load_addr=<u64>`
         // Required: path
         let mut arg_parser = CfgArgParser::new(kernel_cfg_str);
-        let cmdline = arg_parser
+
+        let cmdline_str = arg_parser
             .value_of("cmdline")
             .map_err(ConversionError::new_kernel)?
             .unwrap_or_else(|| DEFAULT_KERNEL_CMDLINE.to_string());
+
+        let mut cmdline = Cmdline::new(KERNEL_CMDLINE_CAPACITY);
+        cmdline
+            .insert_str(cmdline_str)
+            .map_err(|_| ConversionError::new_kernel("Kernel cmdline capacity error"))?;
 
         let path = arg_parser
             .value_of("path")
@@ -258,8 +281,12 @@ mod tests {
     fn test_kernel_config() {
         // Check that additional commas in the kernel string do not cause a panic.
         let kernel_str = r#"path=/foo/bar,cmdline="foo=bar",kernel_load_addr=42,"#;
+
+        let mut foo_cmdline = Cmdline::new(128);
+        foo_cmdline.insert_str("\"foo=bar\"").unwrap();
+
         let expected_kernel_config = KernelConfig {
-            cmdline: r#""foo=bar""#.to_string(),
+            cmdline: foo_cmdline,
             load_addr: 42,
             path: PathBuf::from("/foo/bar"),
         };

--- a/src/vmm/src/lib.rs
+++ b/src/vmm/src/lib.rs
@@ -7,7 +7,6 @@
 #![deny(missing_docs)]
 
 use std::convert::TryFrom;
-use std::ffi::CString;
 use std::fs::File;
 use std::io::{self, stdin, stdout};
 use std::ops::{Deref, DerefMut};
@@ -22,7 +21,7 @@ use kvm_ioctls::{
     Kvm,
 };
 use linux_loader::bootparam::boot_params;
-use linux_loader::cmdline;
+use linux_loader::cmdline::{self, Cmdline};
 use linux_loader::configurator::{
     self, linux::LinuxBootConfigurator, BootConfigurator, BootParams,
 };
@@ -297,7 +296,7 @@ impl TryFrom<VMMConfig> for Vmm {
         // functionality), until we figure out how this fits with the rest of the vmm, which
         // apparently does not explicitly use `Cmdline` structs. Will discuss and fix this
         // somehow ASAP.
-        let mut cmdline = cmdline::Cmdline::new(4096);
+        let mut cmdline = Cmdline::new(4096);
 
         if let Some(cfg) = config.block_config.as_ref() {
             vmm.add_block_device(cfg, &mut cmdline)?;
@@ -401,15 +400,16 @@ impl Vmm {
         bootparams.hdr.cmdline_size = self.kernel_cfg.cmdline.len() as u32 + 1;
 
         // Load the kernel command line into guest memory.
-        // Creating a CString fails when the string contains a 0-byte. In case that happens,
-        // let's just return an InvalidAscii error.
-        let cmdline_str = CString::new(self.kernel_cfg.cmdline.as_str())
-            .map_err(|_| Error::Cmdline(linux_loader::cmdline::Error::InvalidAscii))?;
+        let mut cmdline = Cmdline::new(4096);
+        cmdline
+            .insert_str(self.kernel_cfg.cmdline.as_str())
+            .map_err(Error::Cmdline)?;
+
         load_cmdline(
             &self.guest_memory,
             GuestAddress(CMDLINE_START),
             // Safe because we know the command line string doesn't contain any 0 bytes.
-            &cmdline_str,
+            &cmdline,
         )
         .map_err(Error::KernelLoad)?;
 
@@ -486,11 +486,7 @@ impl Vmm {
     // only support a single device. We need to expand this, but it looks like a good match if we
     // can do it after figuring out how to better separate concerns and make the VMM agnostic of
     // the actual device types.
-    fn add_block_device(
-        &mut self,
-        cfg: &BlockConfig,
-        cmdline: &mut cmdline::Cmdline,
-    ) -> Result<()> {
+    fn add_block_device(&mut self, cfg: &BlockConfig, cmdline: &mut Cmdline) -> Result<()> {
         let mem = Arc::new(self.guest_memory.clone());
 
         let range = MmioRange::new(MmioAddress(MMIO_GAP_START), 0x1000).unwrap();
@@ -521,7 +517,7 @@ impl Vmm {
         Ok(())
     }
 
-    fn add_net_device(&mut self, cfg: &NetConfig, cmdline: &mut cmdline::Cmdline) -> Result<()> {
+    fn add_net_device(&mut self, cfg: &NetConfig, cmdline: &mut Cmdline) -> Result<()> {
         let mem = Arc::new(self.guest_memory.clone());
 
         let range = MmioRange::new(MmioAddress(MMIO_GAP_START + 0x1000), 0x1000).unwrap();
@@ -913,7 +909,7 @@ mod tests {
         let block_config = BlockConfig {
             path: tempfile.as_path().to_path_buf(),
         };
-        let mut cmdline = cmdline::Cmdline::new(4096);
+        let mut cmdline = Cmdline::new(4096);
         assert!(vmm.add_block_device(&block_config, &mut cmdline).is_ok());
         assert_eq!(vmm.block_devices.len(), 1);
         assert!(cmdline.as_str().contains("virtio"));
@@ -923,7 +919,7 @@ mod tests {
             // and delete the underlying file.
             path: TempFile::new().unwrap().as_path().to_path_buf(),
         };
-        let mut cmdline = cmdline::Cmdline::new(4096);
+        let mut cmdline = Cmdline::new(4096);
         let err = vmm
             .add_block_device(&invalid_block_config, &mut cmdline)
             .unwrap_err();
@@ -933,7 +929,7 @@ mod tests {
 
         // The current implementation of the VMM does not allow more than one block device
         // as we are hard coding the `MmioConfig`.
-        let mut cmdline = cmdline::Cmdline::new(4096);
+        let mut cmdline = Cmdline::new(4096);
         // This currently fails because a device is already registered with the provided
         // MMIO range.
         assert!(vmm.add_block_device(&block_config, &mut cmdline).is_err());
@@ -951,7 +947,7 @@ mod tests {
         };
 
         {
-            let mut cmdline = cmdline::Cmdline::new(4096);
+            let mut cmdline = Cmdline::new(4096);
             assert!(vmm.add_net_device(&cfg, &mut cmdline).is_ok());
             assert_eq!(vmm.net_devices.len(), 1);
             assert!(cmdline.as_str().contains("virtio"));
@@ -960,7 +956,7 @@ mod tests {
         {
             // The current implementation of the VMM does not allow more than one net device
             // as we are hard coding the `MmioConfig`.
-            let mut cmdline = cmdline::Cmdline::new(4096);
+            let mut cmdline = Cmdline::new(4096);
             // This currently fails because a device is already registered with the provided
             // MMIO range.
             assert!(vmm.add_net_device(&cfg, &mut cmdline).is_err());

--- a/src/vmm/src/lib.rs
+++ b/src/vmm/src/lib.rs
@@ -290,25 +290,12 @@ impl TryFrom<VMMConfig> for Vmm {
         vmm.add_rtc_device();
 
         // Adding the virtio devices. We'll come up with a cleaner abstraction for `Env`.
-
-        // We transiently define a mut `Cmdline` object here to pass for device creation
-        // (devices expect a `&mut Cmdline` to leverage the newly added virtio_mmio
-        // functionality), until we figure out how this fits with the rest of the vmm, which
-        // apparently does not explicitly use `Cmdline` structs. Will discuss and fix this
-        // somehow ASAP.
-        let mut cmdline = Cmdline::new(4096);
-
         if let Some(cfg) = config.block_config.as_ref() {
-            vmm.add_block_device(cfg, &mut cmdline)?;
+            vmm.add_block_device(cfg)?;
         }
 
         if let Some(cfg) = config.net_config.as_ref() {
-            vmm.add_net_device(cfg, &mut cmdline)?;
-        }
-
-        if !cmdline.as_str().is_empty() {
-            vmm.kernel_cfg.cmdline.push(' ');
-            vmm.kernel_cfg.cmdline.push_str(cmdline.as_str());
+            vmm.add_net_device(cfg)?;
         }
 
         Ok(vmm)
@@ -397,7 +384,7 @@ impl Vmm {
 
         // Add the kernel command line to the boot parameters.
         bootparams.hdr.cmd_line_ptr = CMDLINE_START as u32;
-        bootparams.hdr.cmdline_size = self.kernel_cfg.cmdline.len() as u32 + 1;
+        bootparams.hdr.cmdline_size = self.kernel_cfg.cmdline.as_str().len() as u32 + 1;
 
         // Load the kernel command line into guest memory.
         let mut cmdline = Cmdline::new(4096);
@@ -436,11 +423,16 @@ impl Vmm {
         // See more IRQ assignments & info: https://tldp.org/HOWTO/Serial-HOWTO-8.html
         self.vm.register_irqfd(&interrupt_evt, 4)?;
 
-        self.kernel_cfg.cmdline.push_str(" console=ttyS0");
+        self.kernel_cfg
+            .cmdline
+            .insert_str("console=ttyS0")
+            .map_err(Error::Cmdline)?;
+
         #[cfg(target_arch = "aarch64")]
         self.kernel_cfg
             .cmdline
-            .push_str(&format!(" earlycon=uart,mmio,0x{:08x}", AARCH64_MMIO_BASE));
+            .insert_str(&format!("earlycon=uart,mmio,0x{:08x}", AARCH64_MMIO_BASE))
+            .map_err(Error::Cmdline)?;
 
         // Put it on the bus.
         // Safe to use unwrap() because the device manager is instantiated in new(), there's no
@@ -486,7 +478,7 @@ impl Vmm {
     // only support a single device. We need to expand this, but it looks like a good match if we
     // can do it after figuring out how to better separate concerns and make the VMM agnostic of
     // the actual device types.
-    fn add_block_device(&mut self, cfg: &BlockConfig, cmdline: &mut Cmdline) -> Result<()> {
+    fn add_block_device(&mut self, cfg: &BlockConfig) -> Result<()> {
         let mem = Arc::new(self.guest_memory.clone());
 
         let range = MmioRange::new(MmioAddress(MMIO_GAP_START), 0x1000).unwrap();
@@ -500,7 +492,7 @@ impl Vmm {
             event_mgr: &mut self.event_mgr,
             mmio_mgr: guard.deref_mut(),
             mmio_cfg,
-            kernel_cmdline: cmdline,
+            kernel_cmdline: &mut self.kernel_cfg.cmdline,
         };
 
         let args = BlockArgs {
@@ -517,7 +509,7 @@ impl Vmm {
         Ok(())
     }
 
-    fn add_net_device(&mut self, cfg: &NetConfig, cmdline: &mut Cmdline) -> Result<()> {
+    fn add_net_device(&mut self, cfg: &NetConfig) -> Result<()> {
         let mem = Arc::new(self.guest_memory.clone());
 
         let range = MmioRange::new(MmioAddress(MMIO_GAP_START + 0x1000), 0x1000).unwrap();
@@ -531,7 +523,7 @@ impl Vmm {
             event_mgr: &mut self.event_mgr,
             mmio_mgr: guard.deref_mut(),
             mmio_cfg,
-            kernel_cmdline: cmdline,
+            kernel_cmdline: &mut self.kernel_cfg.cmdline,
         };
 
         let args = NetArgs {
@@ -654,7 +646,7 @@ mod tests {
             kernel_config: KernelConfig {
                 path: default_elf_path(),
                 load_addr: DEFAULT_HIGH_RAM_START,
-                cmdline: DEFAULT_KERNEL_CMDLINE.to_string(),
+                cmdline: KernelConfig::default_cmdline(),
             },
             memory_config: MemoryConfig {
                 size_mib: MEM_SIZE_MIB,
@@ -909,30 +901,27 @@ mod tests {
         let block_config = BlockConfig {
             path: tempfile.as_path().to_path_buf(),
         };
-        let mut cmdline = Cmdline::new(4096);
-        assert!(vmm.add_block_device(&block_config, &mut cmdline).is_ok());
+
+        assert!(vmm.add_block_device(&block_config).is_ok());
         assert_eq!(vmm.block_devices.len(), 1);
-        assert!(cmdline.as_str().contains("virtio"));
+        assert!(vmm.kernel_cfg.cmdline.as_str().contains("virtio"));
 
         let invalid_block_config = BlockConfig {
             // Let's create the tempfile directly here so that it gets out of scope immediately
             // and delete the underlying file.
             path: TempFile::new().unwrap().as_path().to_path_buf(),
         };
-        let mut cmdline = Cmdline::new(4096);
-        let err = vmm
-            .add_block_device(&invalid_block_config, &mut cmdline)
-            .unwrap_err();
+
+        let err = vmm.add_block_device(&invalid_block_config).unwrap_err();
         assert!(
             matches!(err, Error::Block(block::Error::OpenFile(io_err)) if io_err.kind() == ErrorKind::NotFound)
         );
 
         // The current implementation of the VMM does not allow more than one block device
         // as we are hard coding the `MmioConfig`.
-        let mut cmdline = Cmdline::new(4096);
         // This currently fails because a device is already registered with the provided
         // MMIO range.
-        assert!(vmm.add_block_device(&block_config, &mut cmdline).is_err());
+        assert!(vmm.add_block_device(&block_config).is_err());
     }
 
     #[test]
@@ -947,19 +936,17 @@ mod tests {
         };
 
         {
-            let mut cmdline = Cmdline::new(4096);
-            assert!(vmm.add_net_device(&cfg, &mut cmdline).is_ok());
+            assert!(vmm.add_net_device(&cfg).is_ok());
             assert_eq!(vmm.net_devices.len(), 1);
-            assert!(cmdline.as_str().contains("virtio"));
+            assert!(vmm.kernel_cfg.cmdline.as_str().contains("virtio"));
         }
 
         {
             // The current implementation of the VMM does not allow more than one net device
             // as we are hard coding the `MmioConfig`.
-            let mut cmdline = Cmdline::new(4096);
             // This currently fails because a device is already registered with the provided
             // MMIO range.
-            assert!(vmm.add_net_device(&cfg, &mut cmdline).is_err());
+            assert!(vmm.add_net_device(&cfg).is_err());
         }
     }
 }

--- a/src/vmm/tests/integration_tests.rs
+++ b/src/vmm/tests/integration_tests.rs
@@ -17,7 +17,7 @@ fn default_kernel_config(path: PathBuf) -> KernelConfig {
     KernelConfig {
         path,
         load_addr: 0x0010_0000, // 1 MB
-        cmdline: "console=ttyS0 i8042.nokbd reboot=t panic=1 pci=off".to_string(),
+        cmdline: KernelConfig::default_cmdline(),
     }
 }
 

